lib/Seq.dfy

```diff
@@ -287,4 +287,39 @@ module Seq {
   {
   }
 
+  /***** Palindrome *****/
+  predicate palindrome?<T>(s : seq<T>) {
+  if |s| <= 1 then true else (s[0] == s[|s|-1] && palindrome?(s[1..|s|-1]))
+}
+
+lemma palindromeOne<T>(s: seq<T>)
+requires |s| == 1
+ensures palindrome?(s)
+{}
+
+lemma palindromeIndex<T>(s: seq<T>)
+requires palindrome?(s)
+ensures forall i :: 0 <= i < |s| ==> s[i] == s[|s|-1-i]
+{}
+
+
+lemma palindromToReverse<T>(s: seq<T>)
```

Comment:
I was able to prove this lemma by introducing a new variable `var t := s[1..|s| - 1];`:

```dafny
lemma reverseToPalindrome<T>(s: seq<T>)
  requires Reverse(s) == s
  ensures palindrome?(s)
{
  if |s| > 1 {
    var t := s[1..|s| - 1];
    ReverseIndexAll(s);
    ReverseIndexAll(t);
  }
}
```